File: solutions/4095-number-of-balanced-integers-in-a-range/number-of-balanced-integers-in-a-range.ts
function countBalanced(low: number, high: number): number {
    const O = 100;
    const MAXLEN = 20;

    function solve(s: string, tightFirst: boolean): number {
        const n = s.length;
        const memo: number[][][] = Array.from({ length: MAXLEN }, () =>
            Array.from({ length: 200 }, () => [-1, -1])
        );

        function dp(ind: number, bal: number, tight: number): number {
            if (bal < 0 || bal >= 200) return 0;
            if (ind === n) return bal === O ? 1 : 0;
            if (memo[ind][bal][tight] !== -1) return memo[ind][bal][tight];

            let res = 0;
            const lim = tight ? s.charCodeAt(ind) - 48 : 9;

            for (let d = 0; d <= lim; d++) {
                const nt = tight && d === lim ? 1 : 0;
                const nb = bal + (ind % 2 === 0 ? d : -d);
                res += dp(ind + 1, nb, nt);
            }

            memo[ind][bal][tight] = res;
            return res;
        }

        let total = 0;
        const firstLim = tightFirst ? s.charCodeAt(0) - 48 : 9;

        for (let d = 1; d <= firstLim; d++) {
            const tight = tightFirst && d === firstLim ? 1 : 0;
            total += dp(1, O + d, tight);
        }

        return total;
    }

    function len(n: number): number {
        if (n < 10) return 0;
        const s = n.toString();
        let total = 0;

        for (let l = 2; l < s.length; l++) {
            total += solve("9".repeat(l), false);
        }

        total += solve(s, true);
        return total;
    }

    return len(high) - len(low - 1);
}

File: solutions/4095-number-of-balanced-integers-in-a-range/number-of-balanced-integers-in-a-range.go
func countBalanced(low int64, high int64) int64 {
	low -= 1
	dlow := make([]int, 0)
	dhigh := make([]int, 0)

	for low > 0 {
		dlow = append(dlow, int(low%10))
		low /= 10
	}

	for high > 0 {
		dhigh = append(dhigh, int(high%10))
		high /= 10
	}

	diff1 := len(dlow) - len(dhigh)
	diff2 := len(dhigh) - len(dlow)

	for i := 0; i < diff1; i++ {
		dhigh = append(dhigh, 0)
	}
	for i := 0; i < diff2; i++ {
		dlow = append(dlow, 0)
	}

	for i := 0; i < len(dlow)/2; i++ {
		dlow[i], dlow[len(dlow)-1-i] = dlow[len(dlow)-1-i], dlow[i]
		dhigh[i], dhigh[len(dlow)-1-i] = dhigh[len(dlow)-1-i], dhigh[i]
	}

	solve := func(pos, t, odd, even int, digit []int) int64 {
        dp := make([][][][]int64, len(digit))
        for i := 0; i < len(dp); i++ {
            dp[i] = make([][][]int64, 2)
            for j := 0; j < len(dp[i]); j++ {
                dp[i][j] = make([][]int64, 64)
                for k := 0; k < len(dp[i][j]); k++ {
                    dp[i][j][k] = make([]int64, 64)
                    for l := 0; l < len(dp[i][j][k]); l++ {
                        dp[i][j][k][l] = -1
                    }
                }
            }
        }
        
		var dfs func(pos, t, odd, even int, digit []int) int64
		dfs = func(pos, t, odd, even int, digit []int) (res int64) {
            if odd > 63 || even > 63 {return 0}
			if pos == len(digit) && odd == even {
				return 1
			}
			if pos == len(digit) {
				return 0
			}

            if dp[pos][t][odd][even] > -1 {
                return dp[pos][t][odd][even]
            }
            
			l, r := 0, 9
			if t == 1 {
				r = digit[pos]
			}

			for i := l; i <= r; i++ {
				var nt int
				if t == 1 && i == r {
					nt = 1
				}
				var ne, no int = even, odd

				if pos&1 == 0 {
					ne = even + i
				}
				if pos&1 == 1 {
					no = odd + i
				}

				res += dfs(pos+1, nt, no, ne, digit)
			}
			dp[pos][t][odd][even] = res
			return res
		}
		return dfs(0, 1, 0, 0, digit)
	}

	return solve(0, 1, 0, 0, dhigh) - solve(0, 1, 0, 0, dlow)
}
